lib/kbuildbarn/exec/bb_clientd.go

```diff
@@ -75,11 +76,40 @@ func MaybeStartClient(o *ClientOptions) (*Client, error) {
 	if err != nil {
 		return nil, fmt.Errorf("error starting bb_clientd: %w", err)
 	}
+	// From here on out, any returns should return a client handle instead of
+	// nil, so that we can shutdown in case of an error to avoid leaking the
```

Comment:
is this comment correct? well, ok, if ret is declared as a return value. If not, it is just a variable captured by the functor? (can you "name the return value something else/not name it", and just have `if ret != nil && err != nil {
ret.Shutdown()`)?

That way, you still return `nil`, which is more idiomatic for the caller, but the client is still shutdown.
I've also often seen `retErr` simply called `err`.

Reply:
I see what you're saying - creating a client and capturing it in the functor makes the rest of the code less error-prone; switched to that instead.

> I've also often seen retErr simply called err.

With the aliasing that can happen in e.g. `if` blocks, my mind melts. For this reason I like to leave `err` for normal use in the function, and a specifically-named error to be checked in the defer, to disambiguate the two.